operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/CachingFilteringEventSource.java

```diff
@@ -0,0 +1,79 @@
+package io.javaoperatorsdk.operator.processing.event.source;
+
+import java.util.Optional;
+
+import javax.cache.Cache;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.javaoperatorsdk.operator.OperatorException;
+import io.javaoperatorsdk.operator.processing.event.Event;
+import io.javaoperatorsdk.operator.processing.event.ResourceID;
+
+/**
+ * Base class for event sources with filtering and caching capabilities.
+ * <p>
+ * {@link #handleDelete(ResourceID)} - if the related resource is present in the cache it is
+ * removed. and event propagated. There is no event propagated if the resource is not in the cache.
+ * <p>
+ * {@link #handleEvent(Object, ResourceID)} - caches the resource if changed (or nto present
+ * before). Propagates an event if the resource is new or not equals to the one in the cache, and if
+ * accepted by the filter if one is present.
+ *
+ * @param <T> represents the resource (usually external non-kubernetes one) handled.
+ */
+public abstract class CachingFilteringEventSource<T> extends LifecycleAwareEventSource {
+
+  private static final Logger log = LoggerFactory.getLogger(CachingFilteringEventSource.class);
+
+  protected Cache<ResourceID, T> cache;
+  protected EventFilter<T> eventFilter;
+
+  public CachingFilteringEventSource(Cache<ResourceID, T> cache, EventFilter<T> eventFilter) {
+    this.cache = cache;
+    this.eventFilter = eventFilter;
+  }
+
+  protected void handleDelete(ResourceID relatedResourceID) {
+    if (!isRunning()) {
+      log.debug("Received event but event for {} source is not running", relatedResourceID);
+      return;
+    }
+    var cachedValue = cache.get(relatedResourceID);
+    cache.remove(relatedResourceID);
+    // we only propagate event if the resource was previously in cache
+    if (cachedValue != null
+        && (eventFilter == null || eventFilter.acceptDelete(cachedValue, relatedResourceID))) {
```

Comment:
Good question. Wanted to make a generic filter, but this might be over engineered. We can remove the filters for now and see if someone is asking for it.